### PR TITLE
Add backend for Color Kinetics Strips

### PR DESCRIPTION
**Pull request by Firescar96**
Here's an imgur of the lights working: https://imgur.com/a/kdAKJ

tEp also has a panel of LEDs, but they are wired somewhat strangely. I'll have to talk to an alum to see if our setup is normal and if it's worth making a backend for it.

### CODE REVIEW

File: src/KineticsStripOutputNode.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+void KineticsStripOutputNode::onFrame(QSize size, QByteArray frame) {
+  // this is the packet size
+  uint out_packet_idx = 21;
+
+  //convert the rgba values from radiance into rgb for kinetics
+  for (int i = 0; i < frame.size(); i+=4) {
+    out_packet[out_packet_idx++] = frame.at(i)*frame.at(i+3) * 0xFF;
+    out_packet[out_packet_idx++] = frame.at(i+1)*frame.at(i+3) * 0xFF;
+    out_packet[out_packet_idx++] = frame.at(i+2)*frame.at(i+3) * 0xFF;
```
Comment (zbanks) [inline src/KineticsStripOutputNode.cpp]: `frame` uses premultiplied alpha, so you don't need to multiply by alpha again

**Comment by zbanks** [comment]
The build is failing - I think you just need to add `KineticsStripOutputNode.cpp` to https://github.com/zbanks/radiance/blob/master/CMakeLists.txt

File: src/KineticsStripOutputNode.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#include <iostream>
+#include <algorithm>
+
+// Reusable state for sending data packets
+static int out_fd = -1;
+// static struct sockaddr_in out_addr;
+
+// This is dynamic because we need to allocate memory based on
+// the number of pixels per strip.
+unsigned char out_packet[512] = {};
```
Comment (ervanalb) [inline src/KineticsStripOutputNode.cpp]: Can you move these into the class?

You will have to make a `KineticsStripOutputNodePrivate` class and use it to for data storage:

```
class KineticsStripOutputNodePrivate : public SelfTimedReadBackOutputNodePrivate {
public:
    KineticsStripOutputNodePrivate(Context *context); // add additional constructor arguments if you wish
```

Once you do that, you'll also need to create a few boilerplate methods:
```
public:
    KineticsStripOutputNode(const KineticsStripOutputNode &other);
    KineticsStripOutputNode *clone() const override;
private:
    QSharedPointer<KineticsStripOutputNodePrivate> d() const;
```

Take a look at another node type (such as ImageNode) to see what should go into these methods. They are all pretty straightforward.

Sorry that there's so much strange boilerplate, I'm vaguely working on improving it.

File: src/KineticsStripOutputNode.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+std::string ip;
+int startport, width, height;
+struct sockaddr_in out_addr;
```
Comment (ervanalb) [inline src/KineticsStripOutputNode.cpp]: Can you also change these names to be camelCase

File: src/KineticsStripOutputNode.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+KineticsStripOutputNode::KineticsStripOutputNode(Context *context, QSize chainSize)
+: SelfTimedReadBackOutputNode(context, chainSize, 10) {
+
+  connect(this, &SelfTimedReadBackOutputNode::frame, this, &KineticsStripOutputNode::onFrame, Qt::DirectConnection);
+
+  // Try to open a socket
+  if (out_fd > 0) {
+    close(out_fd);
+  }
+  out_fd = socket(AF_INET, SOCK_DGRAM, 0);
```
Comment (ervanalb) [inline src/KineticsStripOutputNode.cpp]: Can you replace these C calls with `QUdpSocket`? http://doc.qt.io/qt-5/qudpsocket.html

File: src/KineticsStripOutputNode.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  ("port", "B", 0x00),
+  ("padding", "B", 0x00),
+  ("flags", "H", 0x0000),
+  ("timer", "I", 0xffffffff),
+  ("universe", "B", 0x00),)*/
+
+  // 21 bytes for the header; 3 bytes (RGB) for each pixel
+  //magic, version, type, sequence, port, padding, flags, timer, universe
+  unsigned char header[21] = {0x04, 0x01, 0xdc, 0x4a, 0x01, 0x00, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff, 0x00};
+  std::copy(header, header + 21, out_packet);
```
Comment (ervanalb) [inline src/KineticsStripOutputNode.cpp]: I would prefer if you used Qt's containers here (probably they are more compatible with `QUdpSocket` as well)

**Comment by Firescar96** [comment]
kk, this is literally my first C++ project so it may take me a bit to process what you said

File: src/KineticsStripOutputNode.cpp (hunk 3 above)
Comment (Firescar96) [reply src/KineticsStripOutputNode.cpp]: whoops, can you tell I ported the code from python?

**Comment by Firescar96** [comment]
I've been struggling for  the past week with threads in C++, time to request assistance. I implemented the  suggestion of using a private class for data storage, but am now getting the error:

```QObject::setParent: Cannot set parent, new parent is in a different thread
[1]    21250 segmentation fault (core dumped)  ./radiance
```
The error is linked to  line 23 of KinectStripOutput node with the call to retrieve a pointer to the private class. The one other class that subclasses SelfTimedReadBackOutputNode does not have any data storage so I can't refer to it for reference.

**Comment by Firescar96** [comment]
Does this look like a bug in the Kinetics output code or something higher up in SelfTimedReadBackOutputNode?

**Comment by Firescar96** [comment]
Both strips and tEpilepsy are working now.
https://imgur.com/a/v9Bl0

I made a modification to SelfTimedReadBackOutputNode which you should comment on. Kept getting an error "Cannot set parent, new parent is in a different thread" without it.

There's probably a more efficient way to write the Kinetics Panel code if I think about it, but this works.

**Comment by Firescar96** [comment]
Note: I'm not sure if you've experienced this, but the effects of running at too high of a framerate had weird effects. 
```
KineticsPanel72OutputNode::KineticsPanel72OutputNode(Context *context, QSize chainSize)
    : SelfTimedReadBackOutputNode(new KineticsPanel72OutputNodePrivate(context, chainSize), context, chainSize, 20) {
```
bumped up the `msec` parameter from 5 to 20 after an event because it caused weird glitches on particular pixels in the light panels. Those pixels would flash, no matter what was being displayed. Not sure whether there is a more definitive way of determining the optimal `msec`.

**Comment by Firescar96** [comment]
fixed the  PR so kinetics nodes are saved on restart

**Comment by ervanalb** [comment]
Hey Nchinda,
This preq has been open for a while and we have since changed the way Radiance outputs to light displays, and I don't think it's practical to merge this in anymore.

You should take a look at https://pypi.org/project/radiance/ and https://github.com/zbanks/radiance/blob/master/light_output.md for how to implement the Color Kinetics lights. With the new method, you shouldn't have to add any C++ code to Radiance.

I think you'll find that your 700 line pull request turns into ~50 lines of Python and will be much easier to deal with. I can definitely help you out if you run into any problems.

**Comment by Firescar96** [comment]
Okay. I have C++ for two Color Kinetics protocols and the Phillips Hue Entertainment API which I'm a bit reluctant to rewrite. So I'll just keep using the current version of Radiance until there are sweet new features I must have.

**Pull request closed without merging**